src/librustc/hir/lowering.rs

```diff
@@ -1734,7 +1742,121 @@ impl<'a> LoweringContext<'a> {
         decl: &FnDecl,
         fn_def_id: Option<DefId>,
         impl_trait_return_allow: bool,
+        fn_ret_transform: &FnRetTransform,
     ) -> P<hir::FnDecl> {
+        // For non async functions, the return type is lowered in a straightforward way
+        impl FnRetTransform for NonAsyncRet {
+            fn lower(&self, ctx: &mut LoweringContext, output: &FunctionRetTy, impl_trait_context: ImplTraitContext) -> hir::FunctionRetTy {
+                match output {
+                    FunctionRetTy::Ty(ty) => hir::Return(ctx.lower_ty(ty, impl_trait_context)),
+                    FunctionRetTy::Default(span) => hir::DefaultReturn(*span),
+                }
+            }
+        }
+
+        // For async closures:
+        //  - If the return type is omitted, a straightforward lowering is performed
+        //  - If the return type is explicit, given a return type T, it is lowered to
+        //          ::std::raw::GenFuture<_, T>
+        impl FnRetTransform for AsyncClosureRet {
+            fn lower(&self, ctx: &mut LoweringContext, output: &FunctionRetTy, impl_trait_context: ImplTraitContext) -> hir::FunctionRetTy {
+                match output {
+                    FunctionRetTy::Ty(ty) => {
+                        let inner_ty = ctx.lower_ty(ty, impl_trait_context);
+                        let span = inner_ty.span;
+                        let hir::Path { def, segments, .. } = ctx.std_path(span, &["raw", "GenFuture"], false);
+                        let LoweredNodeId { node_id, hir_id } = ctx.next_id();
+                        let gen_future_path = hir::Path {
+                            segments: segments.map_slice(|mut v| {
+                                v.last_mut().unwrap().parameters = Some(P(hir::PathParameters {
```

Comment:
Maybe add an `Option<P<hir::PathParameters>>` argument to `std_path` instead?

Reply:
It constructs a `hir::Path` at some point and `map_slice` should be avoidable, yes.

Reply:
It's in the implementation of `resolve_str_path`, turns out: https://github.com/rust-lang/rust/blob/d0456c6f8fe8a88a080b6d4973c8e279de4afe0f/src/librustc_resolve/lib.rs#L1507

I don't know why the path is constructed there instead of its sole caller in `hir::lowering`, but @petrochenkov might.